src/doc/doc_phases/doc_console.nit

```diff
@@ -207,6 +226,189 @@ class CommentQuery
 	end
 end
 
+# A query to search signatures using a specific `MType` as parameter.
+class ParamQuery
+	super MetaQuery
+
+	redef fun perform(nitx, doc) do
+		var res = new Array[NitxMatch]
+		var mtype_name = args.first
+		for mproperty in doc.mproperties do
+			if not mproperty isa MMethod then continue
+			var msignature = mproperty.intro.msignature
+			if msignature != null then
+				for mparam in msignature.mparameters do
+					if mparam.mtype.name == mtype_name then
+						res.add new MEntityMatch(self, mproperty)
+					end
+				end
+			end
+		end
+		return res
+	end
+end
+
+# A query to search signatures using a specific `MType` as return.
+class ReturnQuery
+	super MetaQuery
+
+	redef fun perform(nitx, doc) do
+		var res = new Array[NitxMatch]
+		var mtype_name = args.first
+		for mproperty in doc.mproperties do
+			if not mproperty isa MMethod then continue
+			var msignature = mproperty.intro.msignature
+			if msignature != null then
+				var mreturn = msignature.return_mtype
+				if mreturn != null and mreturn.name == mtype_name then
+					res.add new MEntityMatch(self, mproperty)
+				end
+			end
+		end
+		return res
+	end
+end
+
+# A query to search methods creating new instances of a specific `MType`.
+class NewQuery
+	super MetaQuery
+
+	redef fun perform(nitx, doc) do
+		var res = new Array[NitxMatch]
+		var mtype_name = args.first
+		for mpropdef in doc.mpropdefs do
+			var visitor = new TypeInitVisitor
+			var npropdef = nitx.ctx.modelbuilder.mpropdef2node(mpropdef)
+			if npropdef == null then continue
+			visitor.enter_visit(npropdef)
+			for mtype in visitor.inits do
+				if mtype.name != mtype_name then continue
```

Comment:
Each time you visit and collect all the init then select only those that match the name given.
I suggest that either

* you set the name to search in the visitor so that only useful `new` are collected
* you cache the result of the first visit since, if I understand correctly, future visits will give the same result anyway.